src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts

```diff
@@ -1,87 +1,135 @@
-
-      if (val === "") {
-        if (this._nullable) {
-          return null;
-        }
-        // RPC Quantities must represent `0` as `0x0`
-        return "0x0";
-      }
-      return `0x${val}`;
-    } else if (this.value == null) {
-      return "0x";
-    } else {
-      return super.toString();
+    if (this.bufferValue == null) {
+      return this._nullable? null : Quantity.ZERO_VALUE_STRING;
+    }
+
+    const firstNonZeroByte = this.findFirstNonZeroByteIndex();
+
+    // bufferValue is empty, or contains only 0 bytes
+    if (firstNonZeroByte === this.bufferValue.length) {
+      return Quantity.ZERO_VALUE_STRING;
     }
+
+    let value = this.bufferValue.toString("hex", firstNonZeroByte);
+
+    // only need to check the first char, as we have already skipped 0 bytes in call to this.bufferValue.toString().
+    if (value[0] === "0") {
+      value = value.slice(1);
+    }
+
+    return `0x${value}`;
   }
+
   public toBuffer(): Buffer {
-    // 0x0, 0x00, 0x000, etc should return BUFFER_EMPTY
-    if (Buffer.isBuffer(this.value)) {
-      // trim zeros from start
-      let best = 0;
-      for (best = 0; best < this.value.length; best++) {
-        if (this.value[best] !== 0) break;
-      }
-      if (best > 0) {
-        return this.value.slice(best);
-      } else {
-        return this.value;
-      }
-    } else if (typeof this.value === "string") {
-      let val = this.value.slice(2).replace(/^(?:0+(.+?))?$/, "$1");
-      if (val === "" || val === "0") {
-        return BUFFER_EMPTY;
-      }
-    } else if (this.value === 0 || this.value === 0n) {
+    if (this.bufferValue == null) {
       return BUFFER_EMPTY;
     }
 
-    return super.toBuffer();
-  }
-  public toBigInt(): bigint | null {
-    const value = this.value;
+    const firstNonZeroByte = this.findFirstNonZeroByteIndex();
 
-    // TODO(perf): memoize this stuff
-    if (Buffer.isBuffer(value)) {
-      const bigInt = bufferToBigInt(value);
-      return bigInt == null ? (this._nullable ? null : 0n) : bigInt;
+    if (firstNonZeroByte > 0) {
+      return this.bufferValue.subarray(firstNonZeroByte);
     } else {
-      return value == null ? (this._nullable ? null : 0n) : BigInt(value);
+      return this.bufferValue;
     }
   }
+
+  public toBigInt(): bigint | null {
+    if (this.bufferValue == null) {
+      return this._nullable ? null : 0n;
+    }
+    if (this.bufferValue.length === 0) {
+      return 0n;
+    }
+    return bufferToBigInt(this.bufferValue);
+  }
+
   public toNumber() {
-    // TODO(perf): memoize this stuff
-    return typeof this.value === "number"
-      ? this.value
-      : Number(this.toBigInt());
+    if (this.bufferValue == null) {
+      return this._nullable ? null : 0;
+    }
+
+    const firstNonZeroByte = this.findFirstNonZeroByteIndex();
+
+    const length = this.bufferValue.length - firstNonZeroByte;
+    if (length === 0) {
+      return 0;
+    }
+
+    let result: number;
+    if (length > 6) {
```

Comment:
Why 6? 

Is it worth setting a var for the magic number?

Reply:
I've added a comment, so that the reasoning is more clear (it's more clear than a constant IMO, but open to other POV).

Reply:
Resolving with comment~

I strongly urge adding something like the following to https://github.com/trufflesuite/ganache/blob/develop/src/packages/utils/src/utils/constants.ts

``` javascript
export const BITS_IN_BYTE = 8;
export const MAX_BUFFER_BITS = 48;
export const MAX_BUFFER_BYTES = MAX_BUFFER_BITS / BITS_IN_BYTE; // 6
```

and changing
``` javascript
import { BUFFER_EMPTY, BUFFER_ZERO } from "../../utils/constants";
```

to
``` javascript
import { BUFFER_EMPTY, BUFFER_ZERO, MAX_BUFFER_BYTES } from "../../utils/constants";
```